### ISSUE TITLE
Removing dragged element on hover throws "Expected to find a valid target."

### ISSUE THREAD

**Issue by giladv** [gaearon/react-dnd#361]
hi guys,

i cannot really post a specific code example as it's too cluttered at the moment but i think iv'e isolated the steps required to reproduce the issue.

- start drag
- on hover on dragged item (the item is both drop and drag connected)
   - remove state corresponding to dragged item

the error was thrown at DragDropMonitor.prototype.canDropOnTarget
because it couldn't find a target with the targetID (obviously).

use case: i"m building a sortable tree and i need to relocate the dragged & hovered item
to after it's parent node:
- if it is the last item in a nested list 
- and the hovered item has an horizontal offset to the left
(normal behaviour for tree's: http://pqx.github.io/react-ui-tree/)

on more thing: while the error occurs, it still works.

thanks!

**Comment by rachelsison** [comment]
+1 I am also having this issue

**Comment by gaearon** [comment]
Please provide an example reproducing the issue. I can take a look then.

**Comment by gaearon** [comment]
This is probably fixed in 2.1.0.
I’ll close but please let me know if not, and I’ll reopen.

**Comment by ad1992** [comment]
@gaearon I recently upgraded to react-dnd v5 and this issue is replicated again. Seems like it worked fine in `2.6.0` and also `3.0.2` but breaks in `4.0.5` and `5.0.0`.

**Comment by mwielocha** [comment]
Any news on this?

**Comment by ad1992** [comment]
@gaearon  any update?

**Comment by Jarch09** [comment]
Same problem here

**Comment by drjkuo** [comment]
same here in `5.0.0`

**Comment by happier2** [comment]
same problem here

**Comment by sunndeath** [comment]
same problem here

![image](https://user-images.githubusercontent.com/8881978/51303448-72595200-1a46-11e9-86d9-cee8c058f6ed.png)

**Comment by ravi5b2** [comment]
same problem , Any suggestions for fix ?

**Comment by prateekkarkare** [comment]
Anyone had any luck with this?

**Comment by neilshah2000** [comment]
I have this issue on react-sortable-tree external source sample. Dragging to a red area.

https://frontend-collective.github.io/react-sortable-tree/?path=/story/advanced--drag-from-external-source
https://github.com/frontend-collective/react-sortable-tree/blob/master/stories/external-node.js

`vendors~main.91acd83f50dfbcb80997.bundle.js:1 Uncaught Invariant Violation: Expected to find a valid target.
    at module.exports (https://frontend-collective.github.io/react-sortable-tree/vendors~main.91acd83f50dfbcb80997.bundle.js:2:592)
    at DragDropMonitorImpl.canDropOnTarget (https://frontend-collective.github.io/react-sortable-tree/vendors~main.91acd83f50dfbcb80997.bundle.js:2:186636)
    at DropTargetMonitorImpl.canDrop (https://frontend-collective.github.io/react-sortable-tree/vendors~main.91acd83f50dfbcb80997.bundle.js:2:71541)
    at https://frontend-collective.github.io/react-sortable-tree/main.91acd83f50dfbcb80997.bundle.js:1:132866
    at DragDropContainer.getCurrentState (https://frontend-collective.github.io/react-sortable-tree/vendors~main.91acd83f50dfbcb80997.bundle.js:2:87049)
    at DragDropContainer._this.handleChange (https://frontend-collective.github.io/react-sortable-tree/vendors~main.91acd83f50dfbcb80997.bundle.js:2:84210)
    at handleChange (https://frontend-collective.github.io/react-sortable-tree/vendors~main.91acd83f50dfbcb80997.bundle.js:2:185746)
    at dispatch (https://frontend-collective.github.io/react-sortable-tree/vendors~main.91acd83f50dfbcb80997.bundle.js:2:168058)
    at Object.endDrag (https://frontend-collective.github.io/react-sortable-tree/vendors~main.91acd83f50dfbcb80997.bundle.js:2:199234)
    at HTML5Backend.handleTopDragEndCapture (https://frontend-collective.github.io/react-sortable-tree/vendors~main.91acd83f50dfbcb80997.bundle.js:2:152270)`

**Comment by prateekkarkare** [comment]
The following fix worked for me

https://github.com/frontend-collective/react-sortable-tree/pull/740/commits/9a841136403b219e79c03528f2d70945fca3c44d

**Comment by GNARGNARDAVE** [comment]
I added a debounce to the hover handler (with trailing option) and that solved the issue. The components were updating too quickly by setting the state before DnD could catch up, and the target ID had changed by the time the user dropped the item. So for those of you that could get the above solutions to work, maybe this may help.

**Comment by karlavv** [comment]
> I added a debounce to the hover handler (with trailing option) and that solved the issue. The components were updating too quickly by setting the state before DnD could catch up, and the target ID had changed by the time the user dropped the item. So for those of you that could get the above solutions to work, maybe this may help.

Thank you! @GNARGNARDAVE  That really helped me. I wrapped my dnd function in useCallback and debounce like this: 
`
const moveColumnToDebounce = (dragIndex, hoverIndex) => {
		const dragCol = visibleColumns[dragIndex];
		let arr = visibleColumns.slice();
		arr.splice(dragIndex, 1);
		arr.splice(hoverIndex, 0, dragCol);
		setColumnOrder(arr.map(c => c && c.id));
	};
	const moveColumn = useCallback(debounce(moveColumnToDebounce, 300), [visibleColumns], isEqual);
`
It finally works!

**Comment by GNARGNARDAVE** [comment]
> > I added a debounce to the hover handler (with trailing option) and that solved the issue. The components were updating too quickly by setting the state before DnD could catch up, and the target ID had changed by the time the user dropped the item. So for those of you that could get the above solutions to work, maybe this may help.
> 
> Thank you! @GNARGNARDAVE That really helped me. I wrapped my dnd function in useCallback and debounce like this:
> 
> `const moveColumnToDebounce = (dragIndex, hoverIndex) => {`
> ` const dragCol = visibleColumns[dragIndex];`
> ` let arr = visibleColumns.slice();`
> ` arr.splice(dragIndex, 1);`
> ` arr.splice(hoverIndex, 0, dragCol);`
> ` setColumnOrder(arr.map(c => c && c.id));`
> ` };`
> ` const moveColumn = useCallback(debounce(moveColumnToDebounce, 300), [visibleColumns], isEqual);`
> 
> It finally works!

Glad  that  you were able to  get it to work!

**Comment by TranDangTin** [comment]
> I have this issue on react-sortable-tree external source sample. Dragging to a red area.
> 
> https://frontend-collective.github.io/react-sortable-tree/?path=/story/advanced--drag-from-external-source
> https://github.com/frontend-collective/react-sortable-tree/blob/master/stories/external-node.js
> 
> `vendors~main.91acd83f50dfbcb80997.bundle.js:1 Uncaught Invariant Violation: Expected to find a valid target. at module.exports (https://frontend-collective.github.io/react-sortable-tree/vendors~main.91acd83f50dfbcb80997.bundle.js:2:592) at DragDropMonitorImpl.canDropOnTarget (https://frontend-collective.github.io/react-sortable-tree/vendors~main.91acd83f50dfbcb80997.bundle.js:2:186636) at DropTargetMonitorImpl.canDrop (https://frontend-collective.github.io/react-sortable-tree/vendors~main.91acd83f50dfbcb80997.bundle.js:2:71541) at https://frontend-collective.github.io/react-sortable-tree/main.91acd83f50dfbcb80997.bundle.js:1:132866 at DragDropContainer.getCurrentState (https://frontend-collective.github.io/react-sortable-tree/vendors~main.91acd83f50dfbcb80997.bundle.js:2:87049) at DragDropContainer._this.handleChange (https://frontend-collective.github.io/react-sortable-tree/vendors~main.91acd83f50dfbcb80997.bundle.js:2:84210) at handleChange (https://frontend-collective.github.io/react-sortable-tree/vendors~main.91acd83f50dfbcb80997.bundle.js:2:185746) at dispatch (https://frontend-collective.github.io/react-sortable-tree/vendors~main.91acd83f50dfbcb80997.bundle.js:2:168058) at Object.endDrag (https://frontend-collective.github.io/react-sortable-tree/vendors~main.91acd83f50dfbcb80997.bundle.js:2:199234) at HTML5Backend.handleTopDragEndCapture (https://frontend-collective.github.io/react-sortable-tree/vendors~main.91acd83f50dfbcb80997.bundle.js:2:152270)`

Have you solved this problem yet?

**Comment by scottwilson312** [comment]
I'm having the same issue. It only happens on small lists of 3 or less

**Comment by GNARGNARDAVE** [comment]
I posted a potential solution - if you haven't tried it  yet, I'd suggest putting in a console.log on the hover handler if  you haven't tried to debounce solution on the handler.  From the console log - output the target ids and you'll notice that the target id doesn't  match up because it's triggering too many changes.  Try adding the  debounce to the hover handler

**Comment by ShahStackOverflow** [comment]
let me know if anyone find solution to this problem

**Comment by Aromokeye** [comment]
If you are still facing this problem, check how you update your collection (array) state when the dragging happens. Better to use an immutability helper such as [immerjs](https://immerjs.github.io/immer/) or [immutability-helper](https://github.com/kolodny/immutability-helper).

**Comment by YoannBuzenet** [comment]
Anyone wondering about the Hover handler, here is a working example for me : 

```js
const [{ isOver, canDrop }, drop] = useDrop(
    () => ({
      accept: types.PEC,
      canDrop: () => true,
      drop: () => console.log("it did drop ! On PEC :", order),
      collect: (monitor) => ({
        isOver: !!monitor.isOver(),
        canDrop: !!monitor.canDrop(),
      }),
      hover: debounce(() => setOver(order), 300),
    }),
    []
  );
```

**Comment by alexya** [comment]
Hi, @karlavv  I meet this issue too. Where did u put the function `moveColumn` to? assign it to the `hover` in the useDrop? 
I have a tree structure, when I dragged a node from one node to another node (change the parent), this error will happen. But if I only dragged a node in the current list (don't change the parent, just reorder), it is okay. 
In the hover function, I don't really change the data collection. I updated the data collection just when dropping the node, in `drop` function.
